blog/_posts/2016-03-27-arrays-iteration.md

```diff
@@ -0,0 +1,427 @@
+Bad"](https://en.wikipedia.org/wiki/Alexander_and_the_Terrible,_Horrible,_No_Good,_Very_Bad_Day).
+
+The way we (largely) resolved this problem was to make it possible to
+do cartesian indexing, `A[i,j,k,...]`, for arrays of arbitrary
+dimensionality (the `CartesianIndex` type).  To leverage this in
+practical code, we also had to extend our iterators with the `for I in
+eachindex(A)` construct.  This allows one to select an iterator that
+optimizes the efficiency of access to elements of `A`.  In generic
+algorithms, the performance gains were not small, sometimes on the
+scale of ten- to fifty-fold.  These types were described in a
+[previous blog post](http://julialang.org/blog/2016/02/iteration).
+
+To my knowledge, this approach has given Julia one of the most
+flexible yet efficient "array view" types in any programming language.
+Many languages base views on array *strides*, meaning situations in
+which the memory offset is regular along each dimension.  Among other
+things, this requires that the underlying array is dense.  In
+contrast, in Julia we can easily handle non-strided arrays (e.g.,
+sampling at `[1,3,17,428,...]` along one dimension, or creating a view
+of a `SparseMatrixCSC`).  We can also handle arrays for which there is
+no underlying storage (e.g., `Range`s).  Being able to do this with a
+common infrastructure is part of what makes different optimized array
+types useful in generic programming.
+
+It's also worth pointing out some problems:
+
+- Most importantly, it requires that one adopt a slightly different
+  programming style. Despite being well into another release cycle,
+  this transition is still [not complete, even in Base]
+  (https://github.com/JuliaLang/julia/pull/15434#issuecomment-194991739).
+
+- For algorithms that involve two or more arrays, there's a
+  possibility that their "best" iterators will be of different
+  types. *In principle*, this is a big problem. Consider matrix-vector
+  multiplication, `A[i,j]*v[j]`, where `j` needs to be in-sync for
+  both `A` and `v`, yet you'd also like all of these accesses to be
+  maximally-efficient.  *In practice*, right now this isn't a burning
+  problem: even if our arrays don't all have efficient linear
+  indexing, to my knowledge all of our (dense) array types have
+  efficient cartesian indexing. Since indexing by `N` integers (where
+  `N` is equal to the dimensionality of the array) is always
+  performant, this serves as a reliable default for generic code.
+  (It's worth noting that this isn't true for sparse arrays, and the
+  lack of a corresponding generic solution is probably the main reason
+  we lack a generic API for writing sparse algorithms.)
+
+Unfortunately, I suspect that if we want to add support for certain
+new operations or types (specific examples below), it will force us to
+set the latter problem on fire.
+
+# Challenging examples
+
+Some possible new `AbstractArray` types pose novel challenges.
+
+## `ReshapedArrays` (#15449)
```

Comment:
I suggest making any mention of a GitHub issue a link; there is the one in the title here, and two more in this section.